RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,179 @@
+
+void addMetricValue(<key>, <value>) to add a value for a metric key previously registered.
+
+void addHistogramMetricValue(<key, <value>) to add a value to a histogram type metric key.
+
+The metrics are reported by one of the above reporting flows. The user implementing BaseStatsReporter must adhere to the above Metrics Semantics of Velox while storing the metrics. That implies, the COUNT type metric must not be overridden and continuously grow with time,  On the other hand, the sum metrics which reports the change in aggregated stats over a period of time will overwrite the value against that metric key. Same applies to AVG, RATE and HISTOGRAMs.
+
+At any instant, we have a snapshot of metrics maintained which is overwritten on new updates.
+
+#### How to Export:
+**Prestissimo as exporter:**
+To keep it simple, the prestissimo worker will behave like an exporter and exposes the REST endpoint `/v1/info/health/metrics`. The Prometheus server can be configured to pull metrics from the worker itself. Again, it is up to the user if they want to introduce another layer in between the Time series DB and the worker.
+
+**A sidecar metric exporter:**
+An sidecar HTTP server that fetches metrics from Prestissimo.
+
+### Storing metric values in BaseStatsReporter:
+#### Memory footprint
+We have in total 103 + 28 +  7= 138 metric keys defined in PrestoCPP and Velox. Here is the current split by metric type:
+
+| Metric Type | PrestoCPP Count |Velox Count | Total|
+| ----------- | --------------- |------------|------|
+| COUNT|3| 10| 13   |
+|SUM|19|6|25|
+|AVG|84|1|85|
+|RATE|0|0|0|
+|HISTOGRAM|4|11|15|
+|||Total|138|
+
+We can maintain a mapping of metric key (which can be anywhere between 30 to 50 characters) and metric value in the impelementation of BaseStatsReporter class.
+Each of the above metric value is stored in a 8 byte integer type. At any moment we may have 138 × 8 = 1104 bytes of values data held in-memory. Also, we have 138 × ~50 = 6900 ~ 7K bytes for keys. In total we have 7K + 1104 ~ 8K bytes of memory. Thus holding metrics in memory has least impact on Prestissimo worker.
+Given the above estimates, we decide keep the stats in-memory and we can revisit this if the memory footprint grows.
+
+Note: Estimating the size of a histogram with quantiles is not straight forward, it depends on rotation time and error tolerance settings.
+
+### Metric timestamps
+Since we overwrite metrics in our current design, it is likely that Prometheus may miss updates as it pulls metrics at regular intervals. The timestamps are assigned by Prometheus server (this is recommended) and may not reflect the exact timestamp at which the metric was recorded.
+On other hand, we can maintain a list of metric values and timestamps seen so far in Prestissimo and in the response to the Prometheus we can include this <timestamp, value> pairs. Prometheus must be configured to honor these timestamps. But this approach is not recommended.
```

Comment:
So you are suggesting that the server maintain the time of reporting separately.

Is doing so worthwhile ?

What are the cases in which the reporting and collected timestamps diverge significantly ? If either of them crashes I think they should show up as gaps in the metrics.